stellargraph/core/element_data.py

```diff
@@ -406,13 +434,110 @@ def _create_undirected_adj_lists(self):
             filtered_targets = filtered_targets[:-num_self_loops]
 
         flat_array %= num_edges
-        neigh_counts = np.bincount(self.sources, minlength=self.number_of_nodes)
-        neigh_counts += np.bincount(filtered_targets, minlength=self.number_of_nodes)
-        splits = np.zeros(len(neigh_counts) + 1, dtype=dtype)
-        splits[1:] = np.cumsum(neigh_counts, dtype=dtype)
+        splits = _get_splits(
+            [self.sources, filtered_targets], self.number_of_nodes, dtype
+        )
 
         return FlatAdjacencyList(flat_array, splits)
 
+    def _init_directed_adj_lists_by_other_node_type(self):
+        (
+            self._edges_in_dict_by_other_node_type,
+            self._edges_out_dict_by_other_node_type,
+        ) = self._create_directed_adj_lists_by_other_node_type()
+
+    def _create_directed_adj_lists_by_other_node_type(self):
+        # record the edge ilocs of incoming, outgoing and both-direction edges
+        source_types = self.node_data.type_ilocs[self.sources]
+        target_types = self.node_data.type_ilocs[self.targets]
+
+        def _to_directed_adj_list(arr, other_node_types):
+            index = {}
+
+            # edge ilocs sorted in order of node iloc
+            flat = np.argsort(arr).astype(self._id_index.dtype, copy=False)
+
+            for other_node_type in np.unique(other_node_types):
+
+                # filter node ilocs based on other node type
+                arr_filtered = arr[other_node_types == other_node_type]
+                splits = _get_splits(
+                    [arr_filtered], self.number_of_nodes, self._id_index.dtype
+                )
+
+                # filter edge ilocs based on other node type
+                flat_filtered = flat[other_node_types[flat] == other_node_type]
+
+                index[other_node_type] = FlatAdjacencyList(flat_filtered, splits)
+
+            return index
+
+        return (
+            _to_directed_adj_list(self.targets, source_types),
+            _to_directed_adj_list(self.sources, target_types),
+        )
+
+    def _init_undirected_adj_lists_by_other_node_type(self):
+        self._edges_dict_by_other_node_type = (
+            self._create_undirected_adj_lists_by_other_node_type()
+        )
+
+    def _create_undirected_adj_lists_by_other_node_type(self):
```

Comment:
This function is... impressive. It definitely seems super hard to phrase some of this.